File: app/routers/order_lifecycle.py
"""ORD-011 — order-lifecycle router (transitions, lifecycle read, history, cancel).

Thin HTTP layer over ``app.services.order_lifecycle``. Every handler 503s when
``S.order_lifecycle_enabled`` is off (default). Adjustment / ship-group routes
(ORD-008/009 services) are intentionally NOT mounted here — those services are
out of the lifecycle-core scope and deferred.
"""
from __future__ import annotations

from typing import Any, Dict, List

from fastapi import APIRouter, Depends, HTTPException, Query, Request

from app.auth.deps import AuthenticatedUser
from app.auth.policy import require_admin_or_root_csrf
from app.auth.roles import Role, normalize_role
from app.core.settings import S
from app.models import (
    OrderCancelIn,
    OrderLifecycleOut,
    OrderStatusHistoryEntry,
    OrderTransitionRequest,
    OrderTransitionResult,
)
from app.services.sessions import require_ui_session

router = APIRouter(prefix="/ui/orders", tags=["order-lifecycle"])


def _require_enabled() -> None:
    if not S.order_lifecycle_enabled:
        raise HTTPException(
            status_code=503,
            detail={"code": "feature_disabled", "message": "Order lifecycle is disabled"},
        )


def _require_order_owner_or_admin(order_header: Dict[str, Any], ctx: Dict[str, Any]) -> None:
    role = normalize_role(ctx.get("role"))
    if role in (Role.ADMIN, Role.ROOT):
        return
    if order_header.get("user_id") != ctx.get("user_sub"):
        raise HTTPException(status_code=404, detail={"code": "order_not_found"})


@router.post("/{order_id}/transition", response_model=OrderTransitionResult)
async def transition_order_endpoint(
    order_id: str,
    body: OrderTransitionRequest,
    request: Request,
    user: AuthenticatedUser = Depends(require_admin_or_root_csrf),
) -> OrderTransitionResult:
    _require_enabled()
    from app.services import order_lifecycle

    try:
        return order_lifecycle.transition_order(
            order_id,
            body.target_status.value,
            actor=user.sub,
            reason=body.reason,
            idempotency_key=body.idempotency_key,
        )
    except order_lifecycle.OrderNotFoundError:
        raise HTTPException(404, {"code": "order_not_found"})
    except order_lifecycle.OrderTransitionError as exc:
        raise HTTPException(409, {"code": "illegal_transition", "detail": str(exc)})
    except order_lifecycle.OrderConflictError as exc:
        raise HTTPException(409, {"code": "conflict", "detail": str(exc)})


@router.get("/{order_id}/lifecycle", response_model=OrderLifecycleOut)
async def get_order_lifecycle_endpoint(
    order_id: str,
    include: str = Query(default="", description="Comma-separated: history,adjustments,ship_groups"),
    ctx: Dict[str, Any] = Depends(require_ui_session),
) -> OrderLifecycleOut:
    _require_enabled()
    from app.services import order_lifecycle

    header = order_lifecycle.get_order_header(order_id)
    if header is None:
        raise HTTPException(404, {"code": "order_not_found"})
    _require_order_owner_or_admin(header, ctx)
    include_set = {s.strip() for s in include.split(",") if s.strip()}
    return order_lifecycle.get_order_lifecycle(order_id, include=include_set)


@router.get("/{order_id}/history", response_model=List[OrderStatusHistoryEntry])
async def get_order_history(
    order_id: str,
    ctx: Dict[str, Any] = Depends(require_ui_session),
) -> List[Dict[str, Any]]:
    _require_enabled()
    from app.services import order_lifecycle

    header = order_lifecycle.get_order_header(order_id)
    if header is None:
        raise HTTPException(404, {"code": "order_not_found"})
    _require_order_owner_or_admin(header, ctx)
    return order_lifecycle.list_status_history(order_id)


@router.post("/{order_id}/cancel", response_model=OrderLifecycleOut)
async def cancel_order_endpoint(
    order_id: str,
    body: OrderCancelIn,
    request: Request,
    ctx: Dict[str, Any] = Depends(require_ui_session),
) -> OrderLifecycleOut:
    _require_enabled()
    from app.services import order_lifecycle

    header = order_lifecycle.get_order_header(order_id)
    if header is None:
        raise HTTPException(404, {"code": "order_not_found"})

    role = normalize_role(ctx.get("role"))
    is_admin = role in (Role.ADMIN, Role.ROOT)
    if not is_admin and header.get("user_id") != ctx.get("user_sub"):
        raise HTTPException(404, {"code": "order_not_found"})

    current_status = header.get("lifecycle_status", "")
    owner_cancellable = {"created", "approved"}
    if not is_admin and current_status not in owner_cancellable:
        raise HTTPException(
            409,
            {
                "code": "cancel_not_allowed",
                "detail": f"Owners can only cancel orders in status {sorted(owner_cancellable)}",
            },
        )

    try:
        return order_lifecycle.cancel_order(
            order_id,
            reason=body.reason,
            refund=body.refund,
            actor=ctx["user_sub"],
        )
    except order_lifecycle.OrderTransitionError as exc:
        raise HTTPException(409, {"code": "illegal_transition", "detail": str(exc)})
    except order_lifecycle.OrderConflictError as exc:
        raise HTTPException(409, {"code": "conflict", "detail": str(exc)})

File: frontend/src/api/endpoints/orderLifecycle.ts
import { api } from "@/api/client";

// ─────────────────────────────────────────────────────────────────────────────
// OFBiz ORDER-LIFECYCLE (ORD-001..ORD-015) endpoint wrappers.
//
// Types are defined INLINE here (not in the shared api/types.ts) and mirror the
// LIVE Pydantic models in app/models.py (OrderLifecycleOut, OrderLineItemOut,
// OrderStatusHistoryEntry, OrderTransitionResult, …) and the LIVE router
// app/routers/order_lifecycle.py (prefix /ui/orders).
//
// NOTE: the live router exposes NO "list orders" endpoint. Order discovery is
// driven by entering an order id (lookup-by-id) on the list page; the status
// filter is applied client-side over the looked-up orders.
//
// Feature gate: S.order_lifecycle_enabled (default OFF). When off every handler
// returns 503 {code: "feature_disabled"}. UI handles 503/404 gracefully.
// ─────────────────────────────────────────────────────────────────────────────

export type OrderLifecycleStatus =
  | "created"
  | "approved"
  | "allocated"
  | "picking"
  | "packed"
  | "shipped"
  | "completed"
  | "held"
  | "backorder"
  | "cancelled"
  | "returned";

export type OrderAdjustmentType =
  | "discount"
  | "surcharge"
  | "tax"
  | "shipping"
  | "promotion";

export interface OrderLineItem {
  item_id: string;
  sku: string;
  name: string;
  quantity: number;
  unit_price_cents: number;
  currency: string;
  metadata: Record<string, unknown>;
}

export interface OrderStatusHistoryEntry {
  event_id: string;
  from_status: string | null;
  to_status: string;
  actor: string;
  reason: string;
  ts: number; // Unix seconds
}

export interface OrderAdjustment {
  adjustment_id: string;
  adj_type: OrderAdjustmentType;
  amount_cents: number;
  currency: string;
  label: string;
  taxable: boolean;
  source_rule_ref: string | null;
  created_at: number;
}

export interface ShipGroup {
  ship_group_id: string;
  ship_method: string;
  address_id: string | null;
  item_ids: string[];
  ship_status: OrderLifecycleStatus;
  ship_date_bucket: string | null;
  ship_ts: number | null;
  tracking_number: string | null;
  tracking_url: string | null;
  created_at: number;
  updated_at: number;
}

/** Mirror of app.models.OrderLifecycleOut. */
export interface OrderLifecycle {
  order_id: string;
  status: string; // legacy mirror status
  created_at: string; // ISO string
  updated_at: string; // ISO string
  source_system: string;
  correlation_id: string;
  amount_cents: number;
  currency: string;
  line_item_count: number;
  metadata: Record<string, unknown>;

  // Lifecycle fields (present only when ORDER_LIFECYCLE_ENABLED is on)
  lifecycle_status: OrderLifecycleStatus | null;
  updated_ts: number | null;
  pre_hold_status: string | null;

  // Enriched collections (null = not fetched; [] = fetched but empty)
  adjustments: OrderAdjustment[] | null;
  ship_groups: ShipGroup[] | null;
  status_history: OrderStatusHistoryEntry[] | null;

  // Derived from the transition graph
  allowed_transitions: string[] | null;

  // Joined order items
  line_items: OrderLineItem[] | null;
}

/** Mirror of app.models.OrderTransitionResult. */
export interface OrderTransitionResult {
  order: OrderLifecycle;
  event_id: string;
  from_status: string | null;
  to_status: string;
}

export interface OrderTransitionRequest {
  target_status: OrderLifecycleStatus;
  reason?: string;
  idempotency_key?: string;
}

export interface OrderCancelRequest {
  reason?: string;
  refund?: boolean;
}

/** Comma-separated include groups accepted by GET /lifecycle. */
export type LifecycleInclude = "history" | "adjustments" | "ship_groups";

// ─── List orders (LIVE: GET /ui/orders) ─────────────────────────────────────

/** One row of GET /ui/orders. Owner-scoped for users; admin may filter. */
export interface OrderListItem {
  order_id: string;
  user_id: string;
  status: string;
  lifecycle_status: OrderLifecycleStatus | null;
  created_at: string; // ISO 8601 string (backend OrderListItem.created_at: str)
  updated_at: string; // ISO 8601 string
  source_system: string;
  correlation_id: string;
  amount_cents: number;
  currency: string;
  line_item_count: number;
}

export interface OrderListOut {
  orders: OrderListItem[];
  next_cursor: string | null;
}

/**
 * GET /ui/orders?user_id=&status=&limit=50&cursor=
 * Normal users list their own orders; admins may pass user_id/status filters.
 * Feature-gated: returns 503 {code:"feature_disabled"} when the order-lifecycle
 * flag is off.
 */
export const listOrders = (opts?: {
  userId?: string;
  status?: OrderLifecycleStatus;
  limit?: number;
  cursor?: string;
}) => {
  const params: Record<string, string> = {};
  if (opts?.userId) params["user_id"] = opts.userId;
  if (opts?.status) params["status"] = opts.status;
  params["limit"] = String(opts?.limit ?? 50);
  if (opts?.cursor) params["cursor"] = opts.cursor;
  return api.get<OrderListOut>("/ui/orders", params);
};

// ─── Endpoints (LIVE paths under /ui/orders) ────────────────────────────────

/**
 * GET /ui/orders/{order_id}/lifecycle
 * Enriched order read. `include` requests sub-collections (history/adjustments/
 * ship_groups). Owner-or-admin gated (404 for non-owner non-admin).
 */
export const getOrderLifecycle = (
  orderId: string,
  include: LifecycleInclude[] = [],
) =>
  api.get<OrderLifecycle>(
    `/ui/orders/${encodeURIComponent(orderId)}/lifecycle`,
    include.length ? { include: include.join(",") } : undefined,
  );

/**
 * GET /ui/orders/{order_id}/history
 * Status-history entries, newest-first. Owner-or-admin gated.
 */
export const getOrderHistory = (orderId: string) =>
  api.get<OrderStatusHistoryEntry[]>(
    `/ui/orders/${encodeURIComponent(orderId)}/history`,
  );

/**
 * POST /ui/orders/{order_id}/transition
 * Execute one state-machine transition. ADMIN/ROOT only (CSRF enforced).
 */
export const transitionOrder = (orderId: string, body: OrderTransitionRequest) =>
  api.post<OrderTransitionResult>(
    `/ui/orders/${encodeURIComponent(orderId)}/transition`,
    body,
  );

/**
 * POST /ui/orders/{order_id}/cancel
 * Cancel an order (optionally refund). Owner (only created/approved) or admin.
 */
export const cancelOrder = (orderId: string, body: OrderCancelRequest = {}) =>
  api.post<OrderLifecycle>(
    `/ui/orders/${encodeURIComponent(orderId)}/cancel`,
    body,
  );

// ─── Adjustments ─────────────────────────────────────────────────────────────

export interface AdjustmentListResponse {
  order_id: string;
  adjustments: OrderAdjustment[];
  total_adjustments_cents: number;
}

export interface AddAdjustmentRequest {
  adj_type: OrderAdjustmentType;
  amount_cents: number;
  currency?: string;
  label: string;
  taxable?: boolean;
}

/**
 * GET /ui/orders/{order_id}/adjustments
 * List price adjustments on an order. Owner-or-admin gated.
 */
export const listOrderAdjustments = (orderId: string) =>
  api.get<AdjustmentListResponse>(
    `/ui/orders/${encodeURIComponent(orderId)}/adjustments`,
  );

/**
 * POST /ui/orders/{order_id}/adjustments
 * Add a price adjustment. ADMIN/ROOT only.
 */
export const addOrderAdjustment = (
  orderId: string,
  body: AddAdjustmentRequest,
) =>
  api.post<OrderAdjustment>(
    `/ui/orders/${encodeURIComponent(orderId)}/adjustments`,
    body,
  );

/**
 * DELETE /ui/orders/{order_id}/adjustments/{adj_id}
 * Remove a price adjustment. ADMIN/ROOT only.
 */
export const removeOrderAdjustment = (orderId: string, adjId: string) =>
  api.del<{ ok: boolean }>(
    `/ui/orders/${encodeURIComponent(orderId)}/adjustments/${encodeURIComponent(adjId)}`,
  );

// ─── Ship groups ──────────────────────────────────────────────────────────────

export interface ShipGroupListResponse {
  order_id: string;
  ship_groups: ShipGroup[];
}

export interface CreateShipGroupRequest {
  ship_method: string;
  address_id?: string | null;
  item_ids?: string[];
}

export interface UpdateShipGroupRequest {
  tracking_number?: string | null;
  tracking_url?: string | null;
  ship_status?: OrderLifecycleStatus;
  ship_date_bucket?: string | null;
  ship_ts?: number | null;
}

/**
 * GET /ui/orders/{order_id}/ship-groups
 * List ship groups for an order. Owner-or-admin gated.
 */
export const listShipGroups = (orderId: string) =>
  api.get<ShipGroupListResponse>(
    `/ui/orders/${encodeURIComponent(orderId)}/ship-groups`,
  );

/**
 * POST /ui/orders/{order_id}/ship-groups
 * Create a new ship group. ADMIN/ROOT only.
 */
export const createShipGroup = (
  orderId: string,
  body: CreateShipGroupRequest,
) =>
  api.post<ShipGroup>(
    `/ui/orders/${encodeURIComponent(orderId)}/ship-groups`,
    body,
  );

/**
 * PATCH /ui/orders/{order_id}/ship-groups/{sg_id}
 * Update tracking / status of a ship group. ADMIN/ROOT only.
 */
export const updateShipGroup = (
  orderId: string,
  sgId: string,
  body: UpdateShipGroupRequest,
) =>
  api.patch<ShipGroup>(
    `/ui/orders/${encodeURIComponent(orderId)}/ship-groups/${encodeURIComponent(sgId)}`,
    body,
  );

/**
 * DELETE /ui/orders/{order_id}/ship-groups/{sg_id}
 * Remove a ship group. ADMIN/ROOT only.
 */
export const deleteShipGroup = (orderId: string, sgId: string) =>
  api.del<{ ok: boolean }>(
    `/ui/orders/${encodeURIComponent(orderId)}/ship-groups/${encodeURIComponent(sgId)}`,
  );

// ─── Display helpers ─────────────────────────────────────────────────────────

/** All 11 lifecycle statuses (for filter dropdowns). */
export const ORDER_LIFECYCLE_STATUSES: OrderLifecycleStatus[] = [
  "created",
  "approved",
  "allocated",
  "picking",
  "packed",
  "shipped",
  "completed",
  "held",
  "backorder",
  "cancelled",
  "returned",
];

/** Render integer cents as a currency string. */
export function formatCents(cents: number, currency = "USD"): string {
  try {
    return new Intl.NumberFormat("en-US", {
      style: "currency",
      currency: currency || "USD",
    }).format((cents || 0) / 100);
  } catch {
    return `$${((cents || 0) / 100).toFixed(2)}`;
  }
}

/** Human-friendly title-cased status label. */
export function prettyStatus(status: string | null | undefined): string {
  if (!status) return "—";
  return status.charAt(0).toUpperCase() + status.slice(1);
}
